### DISCUSSION TITLE
Custom exception handler does not provide exception to yielding dependency

### DISCUSSION THREAD

**Discussion by artslob** [fastapi/fastapi#7138] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, Request, Depends
from fastapi.responses import JSONResponse
from starlette import status


class CustomException(Exception):
    pass


async def some_dep():
    print("inside dep")
    try:
        yield 123
    except Exception as e:
        print("exception occurred")  # it disappears with custom handler
    finally:
        print("finally in dep")


def create_app() -> FastAPI:
    app = FastAPI()

    @app.on_event("startup")
    async def startup():
        print("startup action")

    @app.on_event("shutdown")
    async def shutdown():
        print("shutdown action")

    # when commenting line below "exception occurred" line appears on post request
    install_exception_handlers(app)

    @app.post("/qwe")
    async def qwe(a: int = Depends(some_dep)):
        print(a)
        raise CustomException

    return app


def install_exception_handlers(app: FastAPI):
    @app.exception_handler(CustomException)
    async def postgres_error_handler(_: Request, __: CustomException) -> JSONResponse:
        return JSONResponse(
            dict(response="this is json response"),
            status_code=status.HTTP_400_BAD_REQUEST,
        )
```

### Description

* Run FastAPI application
* Call `curl -X POST http://127.0.0.1:8000/qwe` in bash
* Got output:
    ```
    inside dep
    123
    INFO:     127.0.0.1:57400 - "POST /qwe HTTP/1.1" 400 Bad Request
    finally in dep
    ```
* Comment line `install_exception_handlers(app)`
* Reload app
* Call `curl -X POST http://127.0.0.1:8000/qwe` in bash
* Got output (output has `exception occurred` line that should also be with custom exception handler):
    ```
    inside dep
    123
    INFO:     127.0.0.1:57414 - "POST /qwe HTTP/1.1" 500 Internal Server Error
    exception occurred
    finally in dep
    ```

### Environment

* OS: `Linux artslob-laptop 5.4.0-48-generic #52~18.04.1-Ubuntu SMP Thu Sep 10 12:50:22 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`:
* Dependencies:
    ```
    $pip list
    Package            Version
    ------------------ -------
    attrs              20.2.0
    click              7.1.2
    fastapi            0.61.1
    h11                0.11.0
    importlib-metadata 2.0.0
    iniconfig          1.0.1
    packaging          20.4
    pip                20.2.3
    pluggy             0.13.1
    py                 1.9.0
    pydantic           1.6.1
    pyparsing          2.4.7
    pytest             6.1.1
    pytest-asyncio     0.14.0
    setuptools         50.3.0
    six                1.15.0
    starlette          0.13.6
    toml               0.10.1
    typing-extensions  3.7.4.3
    uvicorn            0.12.1
    zipp               3.3.0
    ```

* Python version: 3.7.4

### Additional context

Repo with code: https://github.com/artslob/check-custom-handler-fastapi.

So what we see here is that custom exception handler does not provide context about exception to yielding code in FastAPI dependecy. Its pretty ctitical, because if dependency is database transaction then it can result in false negative session commit (when actually commit is not desired).

**Comment by artslob** [comment, marked as answer]
Ok, looks like its not bug, but a intentional behavior. Its clarified in Starlette docs: https://www.starlette.io/exceptions/.
`HTTPException` is actually _handled exception_, not an _error_. That`s why its not propagated to dependencies.

Looks like a way to add custom handler and also propagate exception as error to dependencies is set handler as default handler for all errors:
```
@app.exception_handler(500)
    async def default_error_handler(_: Request, exc: Exception) -> JSONResponse:
        if isinstance(exc, ValueError):
            ...
        if isinstance(exc, TypeError):
            ...
```
Setting handler for `500` or `Exception` will set it as default handler, and will not mark exceptions as handled.

**Comment by artslob** [comment]
This issue is also duplicate of https://github.com/tiangolo/fastapi/issues/869.

Seems as `HTTPException` and custom error handlers marked as being handled is pretty weird and not obvious behavior for many people.

**Comment by tiangolo** [comment]
Thanks for closing the issue @artslob! ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
